Zero pad chunks before saving

schoentoon opened the pull request:
This way the world is better compressable and backups will be smaller. I made an external tool that does this after the fact and it can make backups of the entire world around 5 to 20% smaller in my testings.

aikar commented on Spigot-Server-Patches/0555-Zero-pad-chunks-before-saving-them-to-the-disk.patch:
```diff
@@ -0,0 +1,27 @@
+             bytebuffer.putInt(0, this.count - 5 + 1);
++
++            // Paper start
++            int remaining = this.count % 4096;
```
this is better served by having a single reused buffer of the entire block size, calculate the last block the "partial" data writes into, and then zero out that block and the remainder BEFORE writing the real data, so we can avoid allocating buffers every write.

This static zero buffer should also be a Direct allocated byte buffer too.

Shevchik replied:
This should also use & 4095 instead. Mod operator is slower than bit and.

aikar commented on Spigot-Server-Patches/0555-Zero-pad-chunks-before-saving-them-to-the-disk.patch:
```diff
@@ -0,0 +1,36 @@
+ 
+     }
++
++    private final byte[] zero = new byte[4096];
```
This needs to be a Direct Memory buffer, and static, as otherwise the JVM copies memory every single write

schoentoon committed:
Zero pad chunks before saving them to the disk

This makes the unused data inside region files more repeatable.
Which will result in better compression in case the entire world directory
is compressed (like in backups). Which will naturally result in smaller backups.

schoentoon commented on Spigot-Server-Patches/0592-Zero-pad-chunks-before-saving-them-to-the-disk.patch:
```diff
@@ -0,0 +1,47 @@
+ 
+         public ChunkBuffer(ChunkCoordIntPair chunkcoordintpair) {
+-            super(8096);
++            super(8192); // Paper
```
I would like to point out that I changed this purely so the `bytebuffer.limit(size)` later on wouldn't throw an exception. As it will throw an exception if the new size is beyond the capacity according to https://docs.oracle.com/javase/8/docs/api/java/nio/Buffer.html#limit-int-
I found that the capacity with the original value would go from 8096 (so too small) to 16192 etc. 8192 probably makes more sense anyway, as it's a multiple of the chunk size of 4096.

aikar replied:
I feel like that's a bug on mojang anyways where they originally had 4096 and wanted to double it and just changed the 4 to an 8 instead.

aikar reviewed (dismissed):
Approving but dev team please test this before merge.

hugmanrique commented on Spigot-Server-Patches/0592-Zero-pad-chunks-before-saving-them-to-the-disk.patch:
```diff
@@ -0,0 +1,47 @@
+             bytebuffer.putInt(0, this.count - 5 + 1);
++
++            // Paper start
++            int size = ((this.count / 4096) + 1) * 4096;
```
If `count` is a multiple of 4096, this yields an incorrect value (+4096 of what's desired).
We need a "`Math.ceilDiv`", which doesn't exist. However, you can prove ceilDiv(m, n)=-floorDiv(-m, n). Maybe Mojang already has a util for this?

Alternatively, albeit slower, you can use `((int) Math.ceil(count / 4096D)) * 4096`

schoentoon replied:
Seems like I incorrectly copied that from one of the other functions. Update incoming where I just use said function instead.

Spottedleaf replied:
int size = (this.count + 4095) & ~4095

limbo-app reviewed (dismissed):
Pull request automatically approved for merging - a core team member has previously approved this pull request.

limbo-app reviewed (approved):
Pull request automatically approved for merging - a core team member has previously approved this pull request.
